### DISCUSSION TITLE
Directus won't run with Plesk / Passenger since 9.15.0

### DISCUSSION THREAD

**Discussion by sapientorius** [directus/directus#15091] [Archive]
### Describe the Bug

Since 9.15.0 Directus does not start properly. 9.14.5 just works as it should.

Plesk is the latest 18.0.45 on Ubuntu 20.0.44 LTS

### To Reproduce

Follow the instructions from the docs for installing Directus in Plesk environment.

After npm install, run app in development mode. Open the URL.

### Errors Shown

Passenger Debug gives me this message:

```
[ D 2022-08-16 00:25:28.9479 36692/T1 age/Spa/SpawnEnvSetupperMain.cpp:747 ]: shellName = 'false' detected as supporting '-l': false
Trace: You required the Node Cluster module, which is incompatible with Passenger, a non-functional shim was returned and your app may still work. However, please remove the related code as soon as possible.
    at Module.require (/usr/share/passenger/helper-scripts/node-loader.js:63:12)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/var/www/vhosts/example.domain/subdomain/node_modules/rate-limiter-flexible/lib/RateLimiterCluster.js:22:17)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1174:10)
    at Module.load (node:internal/modules/cjs/loader:998:32)
    at Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1022:19)
    at Module.require (/usr/share/passenger/helper-scripts/node-loader.js:80:25)
    at require (node:internal/modules/cjs/helpers:102:18)

node:internal/modules/cjs/loader:956
  const err = new Error(message);
              ^

Error: Cannot find module '@directus/app/package.json'
Require stack:
- /var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/resolve-package.js
- /var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/get-extensions.js
- /var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/index.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/extensions.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/services/graphql/index.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/services/index.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/middleware/respond.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/controllers/activity.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/app.js
- /var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/server.js
- /var/www/vhosts/example.domain/subdomain/index.js
- /usr/share/passenger/helper-scripts/node-loader.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:956:15)
    at Function.resolve (node:internal/modules/cjs/helpers:108:19)
    at resolvePackage (/var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/resolve-package.js:9:43)
    at ExtensionManager.getSharedDepsMapping (/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/extensions.js:230:103)
    at ExtensionManager.generateExtensionBundles (/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/extensions.js:202:46)
    at ExtensionManager.load (/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/extensions.js:146:45)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async ExtensionManager.initialize (/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/extensions.js:87:13)
    at async createApp (/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/app.js:108:5)
    at async createServer (/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/server.js:45:38) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/resolve-package.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/get-extensions.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/@directus/shared/dist/cjs/utils/node/index.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/extensions.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/services/graphql/index.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/services/index.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/middleware/respond.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/controllers/activity.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/app.js',
    '/var/www/vhosts/example.domain/subdomain/node_modules/directus/dist/server.js',
    '/var/www/vhosts/example.domain/subdomain/index.js',
    '/usr/share/passenger/helper-scripts/node-loader.js'
  ]
}

```

### What version of Directus are you using?

9.15.0 / 9.15.1

### What version of Node.js are you using?

18.7.0

### What database are you using?

MariaDB 10.3.34

### What browser are you using?

Chrome

### How are you deploying Directus?

Plesk

**Comment by azrikahar** [comment]
> Error: Cannot find module '@directus/app/package.json'

This seems similar to #14706, and was technically fixed by #14728 in 9.15.0, but the fix might be (somehow) incompatible with Plesk 🤔

**Comment by sapientorius** [comment]
I am not sure. I tested different 9.14.x versions and it worked. I also thought it might be due to the pnpm change, but the 9.14.x versions after the change were okay. Maybe the fix broke it for Plesk envorinments.

Edit: Tested it again. 9.14.5 works.

**Comment by rijkvanzanten** [comment]
Moving this to a discussion, as we're discussing running development mode through Plesk, rather than a production build

**Reply by sapientorius** [reply]
It's actually both modes. I just started development mode to be able to easily debug it. But i also occurs in production mode.

**Reply by sapientorius** [reply]
I suggest to convert it back to an issue, because it's not only related to development mode, but a general problem when you try to run it via plesk.

**Comment by Silbad** [comment]
Hello, same problem on Plesk with 9.16.1 :( 
```
(node:373) UnhandledPromiseRejectionWarning: Error: Cannot find module '@directus/app/package.json'
```
Thanks...

**Comment by rijkvanzanten** [comment]
I'm pretty sure this is the offending change: https://github.com/directus/directus/pull/14728/files

That being said, I'm no Plesk user myself, so I don't really know why that would change things. Can you help debug this by checking what `require.main.filename` is in the Plesk environment?

**Reply by lukas-schaetzle** [reply]
`require.main.filename` resolves to `/usr/share/passenger/helper-scripts/node-loader.js`

(see https://github.com/phusion/passenger/blob/stable-6.0/src/helper-scripts/node-loader.js)

**Reply by lukas-schaetzle** [reply]
I can confim that the mentioned change is the cause for the error. Manually editing the js files to the old code lines made it work again.

**Reply by lessborders** [reply]
Reverting those lines + a lot of other lines made it (kinda) work. Now accessing /admin gives 404 errors
![image](https://user-images.githubusercontent.com/100006598/188690864-d674b570-4a75-4f53-98fb-7d34532707fc.png)

**Reply by lukas-schaetzle** [reply]
@lessborders The 404 erros should not result from your hotfix. I've just upgraded to 17.4, applied the hotfix again and there are no errors. For the sake of completeness here are the things you have to change (just search for the commented out lines under the "HOTFIX" title below):

In `node_modules/directus/dist/app.js`:
```
// HOTFIX
// const adminPath = require.resolve('@directus/app', require.main ? { paths: [require.main.filename] } : undefined);
const adminPath = require.resolve('@directus/app');
```

In `node_modules/directus/dist/extensions.js`:
```
// HOTFIX
// const appDir = await fs_extra_1.default.readdir(path_1.default.join((0, node_1.resolvePackage)('@directus/app', (_a = require.main) === null || _a === void 0 ? void 0 : _a.filename), 'dist', 'assets'));
const appDir = await fs_extra_1.default.readdir(path_1.default.join((0, node_1.resolvePackage)('@directus/app'), 'dist', 'assets'));
```

**Reply by paulboudewijn** [reply]
> Seems like the same error as #15091. Would you mind helping with debugging and check what `require.main.filename` resolves to, as mentioned in this comment: [#15091 (comment)](https://github.com/directus/directus/discussions/15091#discussioncomment-3434967)?
> 
> Feel free to answer this question over there to follow up in that same discussion. Perhaps this should/can be accounted in #15498 as a broader fix for such issue, if applicable.

On IIS / IISNode require.main.filename resolves to `C:\Program Files\iisnode\interceptor.js`

**Comment by Silbad** [comment]
@rijkvanzanten Will you release a patch or do you have to edit the files? Thanks

**Reply by rijkvanzanten** [reply]
Once a fix becomes available, yes. As of right now, I don't know what causes this, and it only seems to happen on Plesk specifically 🤔 

That being said, have you tried using the Docker image instead of this npm based approach with Plesk? That should immediately negate any of these environment specific issues

**Reply by Silbad** [reply]
Oh if we could we would have done it already, it's shared hosting, we don't have control over the Plesk interface and therefore over adding the docker extension...

**Comment by Silbad** [comment]
Hello @rijkvanzanten, I try the new version 9.17.0, Plesk always crash...
```
Trace: You required the Node Cluster module, which is incompatible with Passenger, a non-functional shim was returned and your app may still work. However, please remove the related code as soon as possible.
```
and
```
(node:2695) Warning: Accessing non-existent property 'padLevels' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
(node:2695) UnhandledPromiseRejectionWarning: Error: Cannot find module '@directus/app/package.json'
Require stack:
```
Thanks...

**Reply by rijkvanzanten** [reply]
We haven't changed anything around the resolution of the app package, so the same issue would still exist. I am not a Plesk user myself, so could use your help in debugging this.

**Comment by lessborders** [comment]
Problem still exists (Plesk 9.17.1)

**Comment by lukas-schaetzle** [comment]
I've tried some things to identify the app entry file and created a question on [Stack Overflow](https://stackoverflow.com/questions/73565415/identify-entrypoint-file-in-node-js-application-using-passenger) but with no answers so far.

One thing I've just came across is that Passenger injects a JS object with the name `PhusionPassenger`. It looks like this:
```
EventEmitter {
  _events: [Object: null prototype] {},
  _eventsCount: 0,
  _maxListeners: undefined,
  options: {
    UNIX_PATH_MAX: 107,
    api_key: <...>,
    app_env: 'development',
    app_group_name: '<...> (development)',
    app_root: '<...>',
    app_type: 'nodejs',
    base_uri: '/',
    bind_address: '127.0.0.1',
    connect_password: '<...>',
    environment_variables: {
      HTTPS: 'on',
      PASSENGER_COMPILE_NATIVE_SUPPORT_BINARY: '0',
      PASSENGER_DOWNLOAD_NATIVE_SUPPORT_BINARY: '0'
    },
    generic_app: false,
    group: <...>,
    group_uuid: <...>,
    gupid: <...>,
    instance_dir: <...>,
    integration_mode: 'apache',
    load_shell_envvars: true,
    log_file: '',
    log_level: 3,
    node_libdir: '/usr/share/passenger/node',
    passenger_agent_path: '/usr/lib/x86_64-linux-gnu/passenger/support-binaries/PassengerAgent',
    passenger_root: '/usr/share/passenger/phusion_passenger/locations.ini',
    passenger_version: '6.0.13',
    process_title: 'Passenger NodejsApp',
    ruby_libdir: '/usr/share/passenger',
    socket_dir: '/tmp/passenger.gTCwks6/apps.s',
    spawn_method: 'direct',
    start_command: '/opt/plesk/node/16/bin/node /usr/share/passenger/helper-scripts/node-loader.js',
    starts_using_wrapper: true,
    startup_file: '<...>/test.js',
    user: <...>,
    wrapper_supplied_by_third_party: false
  },
  configure: [Function: configure],
  _appInstalled: false,
  [Symbol(kCapture)]: false
}
```

Note the keys `options.start_command` referring to the node-loader file and especially the `options.startup_file` which seems to be the file name normally referred to by `require.main.filename`. So I suggest using a resolve function where we check first whether Directus runs in a Passenger environment with `PhusionPassenger !== undefined` and if yes use the `startup_file` key instead.

Please note that I didn't find any documentation about that object. I've only saw that some people use it in some forum threads about Passenger.

**Reply by lukas-schaetzle** [reply]
See #15498 for a corresponding PR

**Reply by Rar9** [reply]
Any fix for this?

**Comment by nickrum** [comment]
This has been resolved in #15574 which will be included in the next release.

**Reply by Rar9** [reply]
When will it be available?
Npm update is showing nothing

Or is npx check-update also possible?

**Reply by nickrum** [reply]
You'll have to wait for the next release. As we are aiming for a bi-weekly release schedule, it should be available soon.
